### PR TITLE
WIP: Add Release workflow (GitHub actions)

### LINKED ISSUE #137: Github Actions CI: Add deployment
We want to add deployment of the binaries to Github release (draft release should be created).

All together, it should work the same as it works on Travis and Appveyor at the moment.

Good start is to take a look at how we do it in .travis.yml and appveyor.yml and try to replicate that.

@drkameleon was interested in tackling this!

### PR DESCRIPTION
**Pull request by drkameleon**
# Description

Added Release workflow to be executed when a new Tag is created (+ removed create/tags from CI workflow).

Fixes #137 

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## Pending

- [ ] Verify Windows release
- [ ] resolve `cleanpath` dependency for macOS builds
- [ ] Clean up

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        id: create-release
+        uses: actions/create-release@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          tag_name: ${{ github.ref }}
+          release_name: ${{ github.ref }}
+          body: |
+            ${{ github.ref }}
+          draft: false
```
Comment (Martinsos) [inline .github/workflows/release.yaml:93]: We will want this to be true. I prefer being able to inspect that draft and publish it myself, just to make sure all is in order.

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+          mv "$(stack path --local-install-root)/bin/wasp.exe" binary-package/wasp-x86_64.exe
+          mv "$(stack path --project-root)/data" binary-package/data
+          mv tools/run_wasp.ps1 binary-package/wasp.ps1
+        if: matrix.is == 'windows-latest'
+
+      - name: Create release
+        id: create-release
+        uses: actions/create-release@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (Martinsos) [inline .github/workflows/release.yaml:87]: We are using this on multiple places, we might want to put it in global env.

**Review by Martinsos — COMMENTED** [review]
Oho you are already working on this, awesome!
I know it is still in progress so I didn't comment much (just 2 tiny comments).
Main question is: why split it into two workflows? I am guessing you do plan to merge them together? Otherwise we get quite a lot of duplication.

File: .github/workflows/release.yaml (hunk 2 above)
Comment (drkameleon) [reply .github/workflows/release.yaml:87]: good catch

File: .github/workflows/release.yaml (hunk 1 above)
Comment (drkameleon) [reply .github/workflows/release.yaml:93]: I'll fix that ;-)

**Comment by drkameleon** [comment]
> Oho you are already working on this, awesome!
> I know it is still in progress so I didn't comment much (just 2 tiny comments).
> Main question is: why split it into two workflows? I am guessing you do plan to merge them together? Otherwise we get quite a lot of duplication.

Well, since you basically don't make a release every single time you make a commit or a PR, having a separate workflow *does* make sense - that's exactly how I'm doing it with Arturo and it helps keeping everything organized.

Re: duplication. I do get your point. What I've been thinking as an optimization is:

Have the main build work as it does now + create the artifacts (keep the binaries so that they can be reused later).
And then have the 2nd workflow (Release) use these artifacts. So, practically, things will be kept organized and without duplications. I'll just have to find if and how that could be done.

**Comment by Martinsos** [comment]
> > Oho you are already working on this, awesome!
> > I know it is still in progress so I didn't comment much (just 2 tiny comments).
> > Main question is: why split it into two workflows? I am guessing you do plan to merge them together? Otherwise we get quite a lot of duplication.
> 
> Well, since you basically don't make a release every single time you make a commit or a PR, having a separate workflow _does_ make sense - that's exactly how I'm doing it with Arturo and it helps keeping everything organized.
> 
> Re: duplication. I do get your point. What I've been thinking as an optimization is:
> 
> Have the main build work as it does now + create the artifacts (keep the binaries so that they can be reused later).
> And then have the 2nd workflow (Release) use these artifacts. So, practically, things will be kept organized and without duplications. I'll just have to find if and how that could be done.

Ok, makes sense!

Another option might be to put it all in one workflow, and have those release steps be conditional (or even as an another job in the same workflow that depends on build job and is conditional). That might be easier than preserving the artefacts. But, I do get that separating them is somewhat cleaner, regarding responsibilities. Ok hm, let's see how this plays out.

**Comment by drkameleon** [comment]
> > > Oho you are already working on this, awesome!
> > > I know it is still in progress so I didn't comment much (just 2 tiny comments).
> > > Main question is: why split it into two workflows? I am guessing you do plan to merge them together? Otherwise we get quite a lot of duplication.
> > 
> > 
> > Well, since you basically don't make a release every single time you make a commit or a PR, having a separate workflow _does_ make sense - that's exactly how I'm doing it with Arturo and it helps keeping everything organized.
> > Re: duplication. I do get your point. What I've been thinking as an optimization is:
> > Have the main build work as it does now + create the artifacts (keep the binaries so that they can be reused later).
> > And then have the 2nd workflow (Release) use these artifacts. So, practically, things will be kept organized and without duplications. I'll just have to find if and how that could be done.
> 
> Ok, makes sense!
> 
> Another option might be to put it all in one workflow, and have those release steps be conditional (or even as an another job in the same workflow that depends on build job and is conditional). That might be easier than preserving the artefacts. But, I do get that separating them is somewhat cleaner, regarding responsibilities. Ok hm, let's see how this plays out.

I personally like this scenario too. And yes, it should be easier to keep the artifacts. |
Provided that GA gives us the option to conditionally execute the release the jobs (on tag creation). It should I guess...

**Review by Daniellop1 — APPROVED** [review]

**Review by Daniellop1 — APPROVED** [review]

**Comment by Martinsos** [comment]
@drkameleon how are you doing :)? I guess this got lost in the sea of other priorities -> should we try to revive it, get something merged? If there is still a lot to do, maybe we can divide it into smaller parts and merge something now and create an issue for what is left?

**Comment by Martinsos** [comment]
Implemented with https://github.com/wasp-lang/wasp/commit/08c08d30677532024c5d97e24452847cd88ce3af and  https://github.com/wasp-lang/wasp/commit/0d98334c881183bbe305704b17d909e70eb916f3 !

@drkameleon take a look at https://github.com/wasp-lang/wasp/blob/master/.github/workflows/ci.yaml if you are interested, I went with the approach of putting it all in one workflow, to avoid duplication. I think it works very well and is clean, I just needed a couple of `if`s and that was it.

Now we are completely on Github Actions, yay :D!

**Pull request closed without merging**